pkg/mediationcontainer/client_websocket_transport.go

```diff
@@ -35,147 +35,222 @@ func CreateWebSocketConnectionConfig(connConfig *MediationContainerConfig) (*Web
 	return &wsConfig, nil
 }
 
-// ===================================================================================================================
+// ============================ ClientWebSocketTransport - Start, Connect, Close ======================================
 // Implementation of the ITransport for websocket communication to send and receive serialized protobuf message bytes
 type ClientWebSocketTransport struct {
-	ws            *websocket.Conn // created during registration
-	inputStream   chan []byte
-	closeReceived bool
+	ws                       *websocket.Conn // created during Connect()
+	connConfig               *WebSocketConnectionConfig
+	inputStreamCh            chan []byte // unbuffered channel
+	closeRequested           bool
+	stopListenerCh           chan bool // buffered channel
+	connClosedNotificationCh chan bool // channel where the transport connection error will be notified
 }
 
 // Instantiate a new ClientWebsocketTransport endpoint for the client
 // Websocket connection is established with the server
-func CreateClientWebSocketTransport(connConfig *WebSocketConnectionConfig) (*ClientWebSocketTransport, error) {
-	websocket, err := newWebSocketConnection(connConfig) // will be nil if the server is not connected
+func CreateClientWebSocketTransport(connConfig *WebSocketConnectionConfig) *ClientWebSocketTransport {
+	transport := &ClientWebSocketTransport{
+		connConfig:               connConfig,
+	}
+	return transport
+}
 
-	if err != nil {
-		return nil, err
+func (clientTransport *ClientWebSocketTransport) Connect(connClosedNotificationCh chan bool) { // TODO: give an optional timeout to wait for server in performWebSocketConnection()
+	if clientTransport.ws != nil {
+		glog.Infof("[Connect] Closing open websocket before connecting ....")
+		clientTransport.ws.Close()
+		clientTransport.ws = nil
 	}
 
-	transport := &ClientWebSocketTransport{
-		ws:          websocket, //newWebSocketConnection(connConfig),
-		inputStream: make(chan []byte),
+	// loop till server is up or close received
+	clientTransport.performWebSocketConnection() // Blocks or till transport is closed
+
+	if clientTransport.ws == nil {
+		return
 	}
 
+	glog.Infof("[Connect] Connected to server " + clientTransport.GetConnectionId())
+	clientTransport.connClosedNotificationCh = connClosedNotificationCh
+	clientTransport.stopListenerCh = make(chan bool, 1)
```

Comment:
why don't we initialize connClosedNotificationCh and stopListenerCh in Create Method